File: api/src/metrics.js
/**
 * Simple Prometheus metrics endpoint for lks-api-service.
 * Exposes basic HTTP metrics so Prometheus in Oregon can scrape us.
 *
 * Prometheus in us-west-2 reaches this via Inter-Region VPC Peering:
 *   Oregon (10.1.1.x) → pcx-lks-2026 → Virginia (10.0.3.x:9100)
 */

const os = require('os');

// In-memory counters
let requestCount   = 0;
let requestErrors  = 0;
let startTime      = Date.now();

function increment()      { requestCount++; }
function incrementError() { requestErrors++; }

/**
 * Build a Prometheus text-format metrics response.
 * Called by GET /metrics — scraped by Prometheus every 30s.
 */
function buildMetrics() {
  const uptimeSeconds = Math.floor((Date.now() - startTime) / 1000);
  const memUsed       = process.memoryUsage().rss;
  const cpuLoad       = os.loadavg()[0];

  return [
    '# HELP lks_api_requests_total Total HTTP requests handled by lks-api-service',
    '# TYPE lks_api_requests_total counter',
    `lks_api_requests_total{service="lks-api",region="us-east-1"} ${requestCount}`,
    '',
    '# HELP lks_api_errors_total Total HTTP 5xx errors from lks-api-service',
    '# TYPE lks_api_errors_total counter',
    `lks_api_errors_total{service="lks-api",region="us-east-1"} ${requestErrors}`,
    '',
    '# HELP lks_api_uptime_seconds Seconds since lks-api-service started',
    '# TYPE lks_api_uptime_seconds gauge',
    `lks_api_uptime_seconds{service="lks-api",region="us-east-1"} ${uptimeSeconds}`,
    '',
    '# HELP lks_api_memory_bytes_rss Resident memory usage of lks-api-service',
    '# TYPE lks_api_memory_bytes_rss gauge',
    `lks_api_memory_bytes_rss{service="lks-api",region="us-east-1"} ${memUsed}`,
    '',
    '# HELP lks_api_cpu_load_avg1 1-minute CPU load average of the host',
    '# TYPE lks_api_cpu_load_avg1 gauge',
    `lks_api_cpu_load_avg1{service="lks-api",region="us-east-1"} ${cpuLoad.toFixed(4)}`,
    '',
  ].join('\n');
}

module.exports = { increment, incrementError, buildMetrics };

File: api/metrics.py
"""
Prometheus metrics for lks-api-service.
Exposed on a separate Flask app at port 9100 (/metrics).
Scraped by Prometheus in us-west-2 via Inter-Region VPC Peering pcx-lks-2026.
"""

import os
import time
import threading
from flask import Flask, Response

metrics_app = Flask("metrics")

_lock = threading.Lock()
_request_count = 0
_error_count = 0
_start_time = time.time()


def increment_request():
    global _request_count
    with _lock:
        _request_count += 1


def increment_error():
    global _error_count
    with _lock:
        _error_count += 1


def _prom_line(help_text, metric_type, name, labels, value):
    lbl = ",".join(f'{k}="{v}"' for k, v in labels.items())
    return (
        f"# HELP {name} {help_text}\n"
        f"# TYPE {name} {metric_type}\n"
        f"{name}{{{lbl}}} {value}\n"
    )


@metrics_app.get("/metrics")
def metrics():
    uptime = int(time.time() - _start_time)
    region = os.getenv("AWS_REGION", "us-east-1")
    lbl = {"service": "lks-api", "region": region}

    body = "\n".join([
        _prom_line("Total HTTP requests", "counter",
                   "lks_api_requests_total", lbl, _request_count),
        _prom_line("Total HTTP 5xx errors", "counter",
                   "lks_api_errors_total", lbl, _error_count),
        _prom_line("Seconds since service started", "gauge",
                   "lks_api_uptime_seconds", lbl, uptime),
    ])
    return Response(body, mimetype="text/plain; version=0.0.4; charset=utf-8")


@metrics_app.get("/health")
def health():
    return "OK", 200
